eo-runtime/src/main/java/EOorg/EOeolang/EOcage.java

```diff
@@ -108,4 +112,99 @@ public Phi lambda() {
             return new Data.ToPhi(true);
         }
     }
+
+    /**
+     * Class to trace if the cage got into recursion during the dataization.
+     * @since 0.36
+     * @todo #2836:60min Add a new parameter of recursion depth. This parameter
+     *   should be set by user via pom.xml. We can make DATAIZING_CAGES a
+     *   Map and count how many  times the cage was met.
+     * @todo #2836:60min Make the class thread safe. It has private static
+     *  field which can be accessed from differ thread and is not thread safe.
+     *  Needs to synchronize this field.
+     */
+    private static final class PhTracedEnclosure implements Phi {
+
+        /**
+         * Cages that are currently dataizing. If one cage is datazing and
+         * it needs to be dataized inside current dataizing, the cage will be here,
+         */
+        private static final Set<Integer> DATAIZING_CAGES = new HashSet<>();
+
+        /**
+         * Enclosure.
+         */
+        private final Phi enclosure;
+
+        /**
+         * Vertex of cage where the {@link PhTracedEnclosure#enclosure}
+         * was retrieved.
+         */
+        private final int cage;
+
+        /**
+         * Ctor.
+         * @param enclosure Enclosure.
+         * @param cage Vertex of source cage.
+         */
+        PhTracedEnclosure(final Phi enclosure, final int cage) {
+            this.enclosure = enclosure;
+            this.cage = cage;
+        }
+
+        @Override
+        public Phi copy() {
+            return new PhTracedEnclosure(this.enclosure.copy(), this.cage);
+        }
+
+        @Override
+        public String locator() {
+            return this.enclosure.locator();
+        }
+
+        @Override
+        public String forma() {
+            return this.enclosure.forma();
+        }
+
+        @Override
+        public String φTerm() {
+            return this.enclosure.φTerm();
+        }
+
+        @Override
+        public Attr attr(final int pos) {
+            return this.getAttrSafely(() -> this.enclosure.attr(pos));
+        }
+
+        @Override
+        public Attr attr(final String name) {
+            return this.getAttrSafely(() -> this.enclosure.attr(name));
+        }
+
+        @Override
+        public int hashCode() {
+            return this.enclosure.hashCode();
+        }
+
+        @Override
+        public boolean equals(final Object obj) {
+            return obj instanceof Phi && this.hashCode() == obj.hashCode();
+        }
+
+        /**
+         * Get attribute tracing cages.
+         * @param supplier Ordinary way to get attribute.
+         * @return The {@link Attr} if there is no StackOverflow case.
+         */
+        private Attr getAttrSafely(final Supplier<Attr> supplier) {
```

Comment:
@levBagryansky What do you think if we rename this method to `attrTraced` or `tracedAttr`?

Reply:
@volodya-lombrozo `attrTraced` looks like we return traced attribute, but it is not. Maybe `getAttrWithTracingTheCage?`

Reply:
@levBagryansky ok, maybe just `attr` then? Since the signature is different, we can easily overload this method.
I don't like `getAttrWithTracingTheCage` because of [this](https://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html).

Reply:
@levBagryansky [Here](https://www.yegor256.com/2018/08/22/builders-and-manipulators.html) you can read about methods:
> Here is a simple principle for naming methods in OOP, which I’m trying to follow in my code: it’s a verb if it manipulates, it’s a noun if it builds. That’s it. Nothing in between. Methods like saveFile() or getTitle() don’t fit and must be renamed and refactored. 

As for this:
> Ideally, our code should be understandable without documentation.

You are absolutely right.  `attr` - is a method which returns attribute, that is. A reader don't need to know "how" it was created. If I need to know "how" - I will read the method body.

Reply:
@volodya-lombrozo The main sense of the method that it traces. `attr` looks like it just applies its supplier inside. Actually the method does not build, it manipulates.
We use compound name for methods in this projects so it is OK. See https://github.com/objectionary/eo/blob/master/eo-maven-plugin/src/main/java/org/eolang/maven/PlaceMojo.java#L214. 
It would be named `satisfy` because it check if the file satisfies to condition but it would be much worse.

Reply:
@levBagryansky 

> Moreover, methods that “manipulate” must always return void, for example print() or save().

This method returns `attr`, so it's looks like a builder. It doesn't matter "how" exactly you create it and which computations you perform. From a reader perspective it is the method that creates an object.

Reply:
@volodya-lombrozo see how the method is used:
```
        @Override
        public Attr attr(final int pos) {
            return this.getAttrSafely(() -> this.enclosure.attr(pos));
        }
```
`() -> this.enclosure.attr(pos)` creates attribute. The method `getAttrSafely` do it with tracing.

Reply:
@levBagryansky 

> May be not significant how traces but significant that traces.

1. What does the `getAttrSafely` method do?
Answer: It creates an `Attr` object
2. How does the `getAttrSafely` method create an `Attr` object?
Answer: It creates `Attr` object by checking it in a trace set `DATAIZING_CAGES`.

> Here this is important that it traces.

You highlighted this importance in the class name `PhTracedEnclosure`, no need to duplicate it in the method name.

Reply:
@volodya-lombrozo Responsibility of this method is to trace. Creating of attribute is responsibility of `Supplier`.

Reply:
@yegor256 please help

Reply:
@volodya-lombrozo Does such method method clearer describes its responsibility?
```
        private <T> T getItWithTracing(final Supplier<T> supplier) {
            if (EOcage.PhTracedEnclosure.DATAIZING_CAGES.contains(this.cage)) {
                throw new ExFailure("The cage %s is already dataizing", this.cage);
            }
            EOcage.PhTracedEnclosure.DATAIZING_CAGES.add(this.cage);
            final T ret = supplier.get();
            EOcage.PhTracedEnclosure.DATAIZING_CAGES.remove(this.cage);
            return ret;
        }
```

Reply:
@levBagryansky Indeed, this method clearly describes what you mean. From this method I see the following:
1. You use "getIt" to describe that you are retrieving something which you don't know (some general type, let's say.) However  n our code we already know exact types and we can use "attr" instead. In other words, "getIt" - for general method where we don't know types, "attr" - for the specific one, as it is in our case.
2. As for "withTracing" - I already answered it above:
> A reader don't need to know "how" it was created. If I need to know "how" - I will read the method body.

Reply:
@volodya-lombrozo reader don't need to know "how" it traces. If I need to know "how" - I will read the method body. Significant that it traces. 
It is obvious that it gets from its signature. Again, to get is responsibility of supplier. Responsibility of the method is to trace that supplier does. 
Again, the idea of method is in tracing. It can trace by plural ways, and how it traces is written in implementation. Here the name say you that it not just gets it but traces.

Reply:
@yegor256 please help. What option is cleaner?
```java
       @Override
        public Attr attr(final String name) {
            return this.getAttrWithTracing(() -> this.enclosure.attr(name));
        }


```
or
```java
       @Override
        public Attr attr(final String name) {
            return this.attr(() -> this.enclosure.attr(name));
        }
```
There is also an option to make the the callee generic in order to hoghlight its goal:
```java
        @Override
        public Attr attr(final String name) {
            return this.getItWithTracing(() -> this.enclosure.attr(name));
        }

        private <T> T getItWithTracing(final Supplier<T> supplier) {
            if (EOcage.PhTracedEnclosure.DATAIZING_CAGES.contains(this.cage)) {
                throw new ExFailure("The cage %s is already dataizing", this.cage);
            }
            EOcage.PhTracedEnclosure.DATAIZING_CAGES.add(this.cage);
            final T ret = supplier.get();
            EOcage.PhTracedEnclosure.DATAIZING_CAGES.remove(this.cage);
            return ret;
        }
```

Reply:
@yegor256 please check

Reply:
@levBagryansky First, using the word `get` for method names which are not "getters" is an anti-pattern: https://stackoverflow.com/questions/46139501/should-getters-from-domain-classes-not-be-prefixed-with-get-in-ddd-and-why Because of this, `getItWithTracing` is not a good option. 

Second, the `...WithTracing` part is an indicator of temporal coupling between statements inside the method, which is also an anti-pattern: https://blog.ploeh.dk/2011/05/24/DesignSmellTemporalCoupling/

A better and more object-oriented design would be to use... guess what, a decorator. You have a `Supplier` and you want to turn it into another `Supplier`, which does something when its `get()` method is called. Make it this way:

```
@Override
public Attr attr(final String name) {
  return TracingWhileGetting(() -> this.enclosure.attr(name));
}
```

Here, the `TracingWhileGetting` is the `Supplier` that does the "tracing" you need.

Reply:
@yegor256 Looks like [overengineering](https://en.wikipedia.org/wiki/Overengineering) to me. We use this method only in this class and in the single place. I would agree if we used this `TracingWhileGetting` in other `Phi` classes, but we do not do it. Moreover, `Tracing` and `Getting` are participles (a participle [is a verb form](https://en.wikipedia.org/wiki/Participle).) But according with [the Java specification](https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html#:~:text=Class%20names%20should%20be%20nouns,such%20as%20URL%20or%20HTML):
> Class names should be nouns, in mixed case with the first letter of each internal word capitalized. 

Nouns (not adjectives or verbs.)  Maybe it's better to name the method `attr`? In this case we won't add extra code and will avoid the all convention violations which you mentioned before. What do you think?